File: tests/learning_rate_check.py
import numpy as np
import xgboost as xgb
import shap

# --- data ---
np.random.seed(0)
n, p = 200, 5
X = np.random.randn(n, p)
y = 3*X[:,0] - 2*X[:,1] + 0.5*X[:,2] + np.random.randn(n)*0.1

dX = xgb.DMatrix(X, label=y)

def check_eta(eta):
    params = {
        "objective": "reg:squarederror",
        "max_depth": 3,
        "eta": eta,
        "seed": 0,
    }
    model = xgb.train(params, dX, num_boost_round=20)

    # 1) global additivity check (SHAP + base == predict)
    expl = shap.TreeExplainer(model)
    shap_vals = expl.shap_values(X)               # (n, p)
    base = expl.expected_value                    # scalar for regression
    pred = model.predict(dX)                      # (n,)

    err = np.max(np.abs(shap_vals.sum(1) + base - pred))

    # 2) marginal tree SHAP magnitude check (tree 2 - tree 1)
    c1 = model.predict(dX, pred_contribs=True, iteration_range=(0, 1))
    c2 = model.predict(dX, pred_contribs=True, iteration_range=(0, 2))
    marg2 = c2 - c1
    marg_mean = np.mean(np.abs(marg2[:, :p]))

    return model, err, marg_mean

for eta in [1.0, 0.3, 0.1]:
    model, err, marg_mean = check_eta(eta)
    print(f"eta={eta:0.1f}  additivity_max_err={err:.2e}  mean(|marg_tree2_shap|)={marg_mean:.4f}")

File: qshapr.Rcheck/00_pkg_src/qshapr/tests/learning_rate_check.R
library(xgboost)

set.seed(0)
n <- 200; p <- 5
X <- matrix(rnorm(n*p), n, p)
y <- 3*X[,1] - 2*X[,2] + 0.5*X[,3] + rnorm(n, sd=0.1)

dtrain <- xgb.DMatrix(X, label = y)

check_eta <- function(eta, nrounds = 20, max_depth = 3) {
  params <- list(
    objective = "reg:squarederror",
    max_depth = max_depth,
    eta = eta,
    seed = 0
  )

  model <- xgb.train(params = params, data = dtrain, nrounds = nrounds, verbose = 0)

  # 1) additivity check: contribs (incl. BIAS) sums to prediction
  contrib_all <- predict(model, X, predcontrib = TRUE)  # n x (p+1), last col is BIAS
  pred <- predict(model, X)
  add_err <- max(abs(rowSums(contrib_all) - pred))

  # 2) marginal SHAP for tree 2 (round 2): contrib(1..2) - contrib(1..1)
  c1 <- predict(model, X, predcontrib = TRUE, iterationrange = c(1, 2))  # round 1 only
  c2 <- predict(model, X, predcontrib = TRUE, iterationrange = c(1, 3))  # rounds 1..2
  marg2 <- c2 - c1

  # mean abs marginal over features only (exclude BIAS)
  marg_mean <- mean(abs(marg2[, 1:p, drop = FALSE]))

  list(model = model, additivity_max_err = add_err, mean_abs_marg_tree2 = marg_mean)
}

for (eta in c(1.0, 0.3, 0.1)) {
  out <- check_eta(eta)
  cat(sprintf("eta=%.1f  additivity_max_err=%.2e  mean(|marg_tree2_shap|)=%.4f\n",
              eta, out$additivity_max_err, out$mean_abs_marg_tree2))
}
